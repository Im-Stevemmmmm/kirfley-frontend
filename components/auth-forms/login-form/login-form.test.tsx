import { MockedProvider, MockedResponse } from "@apollo/client/testing";
import { fireEvent, render } from "@testing-library/react";
import {
    LoginDocument,
    LoginMutation,
    LoginMutationVariables,
} from "generated/graphql-types";
import LoginForm from "./login-form";

test("renders correctly", () => {
    const { container } = render(
        <MockedProvider>
            <LoginForm />
        </MockedProvider>
    );

    expect(container).toMatchSnapshot();
});

test("username or email input is displayed", () => {
    const { getByLabelText } = render(
        <MockedProvider>
            <LoginForm />
        </MockedProvider>
    );
    const usernameInput = getByLabelText(/Username or Email/i);

    expect(usernameInput).toBeInTheDocument;
});

test("password input is displayed", () => {
    const { getByLabelText } = render(
        <MockedProvider>
            <LoginForm />
        </MockedProvider>
    );
    const passwordInput = getByLabelText(/Password/i);

    expect(passwordInput).toBeInTheDocument;
});

test("logs user in on submit and redirects to /home", () => {
    const useRouter = jest.spyOn(require("next/router"), "useRouter");

    useRouter.mockImplementation(() => ({
        route: "",
        basePath: "",
        pathname: "e",
        query: {},
        asPath: "",
        push: async () => true,
        replace: async () => true,
        reload: () => null,
        back: () => null,
        prefetch: async () => undefined,
        beforePopState: () => null,
        isFallback: false,
        events: {
            on: () => null,
            off: () => null,
            emit: () => null,
        },
    }));

    const usernameOrEmail = "Steve";
    const password = "12345";

    const apiMocks: MockedResponse<Record<string, any>>[] = [
        {
            request: {
                query: LoginDocument,
                variables: {
                    usernameOrEmail,
                    password,
                } as LoginMutationVariables,
            },
            result: {
                data: {
                    login: {
                        successful: true,
                    },
                } as LoginMutation,
            },
        },
    ];

    const { getByText, getByLabelText } = render(
        <MockedProvider mocks={apiMocks}>
            <LoginForm />
        </MockedProvider>
    );
    const usernameOrEmailInput = getByLabelText(/Username or Email/i);
    const passwordInput = getByLabelText(/Password/i);
    const submitButton = getByText(/Log in/i, { selector: "button" });

    fireEvent.change(usernameOrEmailInput, {
        target: { value: usernameOrEmail },
    });
    fireEvent.change(passwordInput, { target: { value: password } });

    fireEvent.click(submitButton);

    expect(useRouter).toHaveBeenCalled();
});
