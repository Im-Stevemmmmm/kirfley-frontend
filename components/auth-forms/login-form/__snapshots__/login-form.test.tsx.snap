// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <form
    class="form"
  >
    <h1
      class="form__title"
    >
      Log in
    </h1>
    <label
      class="form__label"
      for="usernameOrEmail"
    >
      Username or Email
    </label>
    <input
      class="form__input"
      id="usernameOrEmail"
      name="usernameOrEmail"
      placeholder="Username or email"
      type="text"
      value=""
    />
    <label
      class="form__label"
      for="password"
    >
      Password
    </label>
    <input
      class="form__input"
      id="password"
      name="password"
      placeholder="Password"
      type="password"
      value=""
    />
    <div
      class="buttonContainer"
    >
      <button
        class="form__button standard"
        type="submit"
      >
        Log In
      </button>
      <p>
        or
      </p>
      <button
        class="form__button standard"
        type="button"
      >
        Create an account
      </button>
    </div>
  </form>
</div>
`;
